Window function got mismatch data type pack error in disaggregated tiflash mode

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1. Setup **disaggregated tiflash** cluster using `tiup cluster`(1 CN 1 WN)
2. Connect to tidb, create table and insert table. [tbl_8.txt](https://github.com/pingcap/tiflash/files/11394498/tbl_8.txt)
3. run the following sqls:
```
alter table tbl_8 set tiflash replica 1;
set tidb_isolation_read_engines = 'tiflash';
select    lag( tbl_8.col_43 , 6 ,NULL) over w as r0 from tbl_8 where not( tbl_8.col_44 between '08:17:52.00' and '05:02:37.00' )  window w as (  order by col_44,col_42,col_40,col_41,col_45,col_43 rows between 2 preceding and current    row ) order by r0 ;
```
<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)
run ok
### 3. What did you see instead (Required)
```
[2023/05/04 14:23:49.021 +08:00] [ERROR] [MPPTask.cpp:471] ["task running meets error: Code: 48, e.displayText() = DB::Exception: Reading mismatch data type pack. Cast from Int64 to MyDuration(0) is NOT supported!: (while reading from DTFile: s3://s4/data/t_154/dmf_19), e.what() = DB::Exception, Stack trace:
       0x1c1608e\tDB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) [tiflash+29450382]
                \tdbms/src/Common/Exception.h:46
       0x74d1ad4\tDB::DM::convertColumnByColumnDefineIfNeed(std::__1::shared_ptr<DB::IDataType const> const&, COWPtr<DB::IColumn>::immutable_ptr<DB::IColumn>&&, DB::DM::ColumnDefine const&) [tiflash+122493652]
                \tdbms/src/Storages/DeltaMerge/convertColumnTypeHelpers.cpp:464
       0x74c0d0c\tDB::DM::DMFileReader::read() [tiflash+122424588]
                \tdbms/src/Storages/DeltaMerge/File/DMFileReader.cpp:625
       0x74b8145\tDB::DM::DMFileBlockInputStream::read() [tiflash+122388805]
                \tdbms/src/Storages/DeltaMerge/File/DMFileBlockInputStream.h:64
       0x74258bd\tDB::DM::ConcatSkippableBlockInputStream<false>::read() [tiflash+121788605]
                \tdbms/src/Storages/DeltaMerge/SkippableBlockInputStream.h:175
       0x744c371\tDB::DM::readBlock(std::__1::shared_ptr<DB::DM::SkippableBlockInputStream>&, std::__1::shared_ptr<DB::DM::SkippableBlockInputStream>&) [tiflash+121946993]
                \tdbms/src/Storages/DeltaMerge/ReadUtil.cpp:33
       0x747c0dc\tDB::DM::BitmapFilterBlockInputStream::readImpl(DB::PODArray<unsigned char, 4096ul, Allocator<false>, 15ul, 16ul>*&, bool) [tiflash+122142940]
                \tdbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilterBlockInputStream.cpp:40
       0x747c6d4\tDB::DM::BitmapFilterBlockInputStream::readImpl() [tiflash+122144468]
                \tdbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilterBlockInputStream.h:46
       0x7667285\tDB::IProfilingBlockInputStream::read(DB::PODArray<unsigned char, 4096ul, Allocator<false>, 15ul, 16ul>*&, bool) [tiflash+124154501]
                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:75
       0x821ead7\tDB::DM::RNRemoteSegmentThreadInputStream::readImpl(DB::PODArray<unsigned char, 4096ul, Allocator<false>, 15ul, 16ul>*&, bool) [tiflash+136440535]
                \tdbms/src/Storages/DeltaMerge/Remote/RNRemoteSegmentThreadInputStream.cpp:147
       0x821f944\tDB::DM::RNRemoteSegmentThreadInputStream::readImpl() [tiflash+136444228]
                \tdbms/src/Storages/DeltaMerge/Remote/RNRemoteSegmentThreadInputStream.h:80
       0x7667285\tDB::IProfilingBlockInputStream::read(DB::PODArray<unsigned char, 4096ul, Allocator<false>, 15ul, 16ul>*&, bool) [tiflash+124154501]
                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:75
       0x7666f75\tDB::IProfilingBlockInputStream::read() [tiflash+124153717]
                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:43
       0x792642d\tDB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, false>::Handler, (DB::StreamUnionMode)0>::loop(unsigned long, DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, false>::Handler, (DB::StreamUnionMode)0>::WorkingInputs&) [tiflash+127034413]
                \tdbms/src/DataStreams/ParallelInputsProcessor.h:307
       0x792628d\tDB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, false>::Handler, (DB::StreamUnionMode)0>::work(unsigned long, DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, false>::Handler, (DB::StreamUnionMode)0>::WorkingInputs&) [tiflash+127033997]
                \tdbms/src/DataStreams/ParallelInputsProcessor.h:270
       0x7926202\tstd::__1::__function::__func<DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, false>::Handler, (DB::StreamUnionMode)0>::process()::'lambda'(), std::__1::allocator<DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, false>::Handler, (DB::StreamUnionMode)0>::process()::'lambda'()>, void ()>::operator()() [tiflash+127033858]
                \t/usr/local/bin/../include/c++/v1/__functional/function.h:345
       0x1d1de38\tauto DB::wrapInvocable<std::__1::function<void ()> >(bool, std::__1::function<void ()>&&)::'lambda'()::operator()() [tiflash+30531128]
                \tdbms/src/Common/wrapInvocable.h:36
       0x1d1df85\tstd::__1::packaged_task<void ()>::operator()() [tiflash+30531461]
                \t/usr/local/bin/../include/c++/v1/future:2089
       0x1c25fa3\tDB::DynamicThreadPool::executeTask(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&) [tiflash+29515683]
                \tdbms/src/Common/DynamicThreadPool.cpp:101
       0x1c25604\tDB::DynamicThreadPool::fixedWork(unsigned long) [tiflash+29513220]
                \tdbms/src/Common/DynamicThreadPool.cpp:115
       0x1c26b12\tvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(unsigned long), DB::DynamicThreadPool*, unsigned long&>(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(unsigned long), DB::DynamicThreadPool*&&, unsigned long&)::'lambda'(auto&&...), DB::DynamicThreadPool*, unsigned long> >(void*) [tiflash+29518610]
                \t/usr/local/bin/../include/c++/v1/thread:291
  0x7effabe46802\tstart_thread [libc.so.6+653314]
  0x7effabde6314\t__GI___clone [libc.so.6+258836]"] [source="MPP<query:<query_ts:1683181428993135345, local_query_id:32, server_id:3262309, start_ts:441235912519057409>,task_id:1>"] [thread_id=235]
```
### 4. What is your TiFlash version? (Required)
```
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                           |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v7.2.0-alpha
Edition: Community
Git Commit Hash: 7eff8ce629d00cf4ab1eeda82a1ecd5a4e394b02
Git Branch: heads/refs/tags/v7.2.0-alpha
UTC Build Time: 2023-05-03 14:33:44
GoVersion: go1.20.3
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: tikv |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.001 sec)
```

--- a/dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp
+++ b/dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp
@@ -96,7 +96,6 @@ std::tuple<DM::ColumnDefinesPtr, size_t> genColumnDefinesForDisaggregatedRead(co
         // Even if the id is pk_column or extra_table_id, we still output it as
         // a exchange receiver output column
         const auto output_name = genNameForExchangeReceiver(i);
-        // TODO: Support generated column, which is a virtual column.
         switch (column_info.id)
         {
         case TiDBPkColumnID:

--- a/dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp
+++ b/dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp
@@ -85,14 +85,14 @@ NamesAndTypes genNamesAndTypes(const TiDBTableScan & table_scan, const StringRef
 
 std::tuple<DM::ColumnDefinesPtr, size_t> genColumnDefinesForDisaggregatedRead(const TiDBTableScan & table_scan)
 {
-    // Now the upper level seems treat disagg read as an ExchangeReceiver output, so
-    // use this as output column prefix.
     auto column_defines = std::make_shared<DM::ColumnDefines>();
     size_t extra_table_id_index = InvalidColumnID;
     column_defines->reserve(table_scan.getColumnSize());
     for (Int32 i = 0; i < table_scan.getColumnSize(); ++i)
     {
         const auto & column_info = table_scan.getColumns()[i];
+        // Now the upper level seems treat disagg read as an ExchangeReceiver output, so
+        // use this as output column prefix.
         // Even if the id is pk_column or extra_table_id, we still output it as
         // a exchange receiver output column
         const auto output_name = genNameForExchangeReceiver(i);

--- a/dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp
+++ b/dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp
@@ -54,7 +54,7 @@ NamesAndTypes genNamesAndTypesForExchangeReceiver(const TiDBTableScan & table_sc
 
 String genNameForExchangeReceiver(Int32 col_index)
 {
-    return "exchange_receiver_" + std::to_string(col_index);
+    return fmt::format("exchange_receiver_{}", col_index);
 }
 
 NamesAndTypes genNamesAndTypes(const ColumnInfos & column_infos, const StringRef & column_prefix)
@@ -87,7 +87,6 @@ std::tuple<DM::ColumnDefinesPtr, size_t> genColumnDefinesForDisaggregatedRead(co
 {
     // Now the upper level seems treat disagg read as an ExchangeReceiver output, so
     // use this as output column prefix.
-    constexpr std::string_view column_prefix = "exchange_receiver";
     auto column_defines = std::make_shared<DM::ColumnDefines>();
     size_t extra_table_id_index = InvalidColumnID;
     column_defines->reserve(table_scan.getColumnSize());
@@ -96,7 +95,8 @@ std::tuple<DM::ColumnDefinesPtr, size_t> genColumnDefinesForDisaggregatedRead(co
         const auto & column_info = table_scan.getColumns()[i];
         // Even if the id is pk_column or extra_table_id, we still output it as
         // a exchange receiver output column
-        const auto output_name = fmt::format("{}_{}", column_prefix, i);
+        const auto output_name = genNameForExchangeReceiver(i);
+        // TODO: Support generated column, which is a virtual column.
         switch (column_info.id)
         {
         case TiDBPkColumnID:
@@ -119,7 +119,7 @@ std::tuple<DM::ColumnDefinesPtr, size_t> genColumnDefinesForDisaggregatedRead(co
             column_defines->emplace_back(DM::ColumnDefine{
                 column_info.id,
                 output_name,
-                getDataTypeByColumnInfoForComputingLayer(column_info),
+                getDataTypeByColumnInfoForDisaggregatedStorageLayer(column_info),
                 column_info.defaultValueToField()});
             break;
         }

--- a/dbms/src/Storages/StorageDisaggregated.cpp
+++ b/dbms/src/Storages/StorageDisaggregated.cpp
@@ -47,8 +47,7 @@ BlockInputStreams StorageDisaggregated::read(
     size_t,
     unsigned num_streams)
 {
-    /// S3 config is enabled on the TiFlash compute node, let's read
-    /// data from S3.
+    /// S3 config is enabled on the TiFlash compute node, let's read data from S3.
     bool remote_data_read = S3::ClientFactory::instance().isEnabled();
     if (remote_data_read)
         return readThroughS3(db_context, query_info, num_streams);

--- a/dbms/src/Storages/StorageDisaggregated.cpp
+++ b/dbms/src/Storages/StorageDisaggregated.cpp
@@ -51,10 +51,14 @@ BlockInputStreams StorageDisaggregated::read(
     /// data from S3.
     bool remote_data_read = S3::ClientFactory::instance().isEnabled();
     if (remote_data_read)
-        return readFromWriteNode(db_context, query_info, num_streams);
+        return readThroughS3(db_context, query_info, num_streams);
 
     /// Fetch all data from write node through MPP exchange sender/receiver
+    return readThroughExchange(num_streams);
+}
 
+BlockInputStreams StorageDisaggregated::readThroughExchange(unsigned num_streams)
+{
     auto remote_table_ranges = buildRemoteTableRanges();
 
     // only send to tiflash node with label {"engine": "tiflash"}
@@ -71,7 +75,9 @@ BlockInputStreams StorageDisaggregated::read(
 
     NamesAndTypes source_columns = genNamesAndTypesForExchangeReceiver(table_scan);
     assert(exchange_receiver->getOutputSchema().size() == source_columns.size());
-    filterConditions(std::move(source_columns), pipeline);
+    analyzer = std::make_unique<DAGExpressionAnalyzer>(std::move(source_columns), context);
+    // TODO: push down filter conditions to write node
+    filterConditions(*analyzer, pipeline);
 
     return pipeline.streams;
 }
@@ -289,14 +295,12 @@ void StorageDisaggregated::buildReceiverStreams(const std::vector<RequestAndRegi
     });
 }
 
-void StorageDisaggregated::filterConditions(NamesAndTypes && source_columns, DAGPipeline & pipeline)
+void StorageDisaggregated::filterConditions(DAGExpressionAnalyzer & analyzer, DAGPipeline & pipeline)
 {
-    analyzer = std::make_unique<DAGExpressionAnalyzer>(std::move(source_columns), context);
-
     if (filter_conditions.hasValue())
     {
         // No need to cast, because already done by tiflash_storage node.
-        ::DB::executePushedDownFilter(/*remote_read_streams_start_index=*/pipeline.streams.size(), filter_conditions, *analyzer, log, pipeline);
+        ::DB::executePushedDownFilter(/*remote_read_streams_start_index=*/pipeline.streams.size(), filter_conditions, analyzer, log, pipeline);
 
         auto & profile_streams = context.getDAGContext()->getProfileStreamsMap()[filter_conditions.executor_id];
         pipeline.transform([&profile_streams](auto & stream) { profile_streams.push_back(stream); });

--- a/dbms/src/Storages/StorageDisaggregated.h
+++ b/dbms/src/Storages/StorageDisaggregated.h
@@ -84,10 +84,12 @@ public:
 
 private:
     // helper functions for building the task read from a shared remote storage system (e.g. S3)
-    BlockInputStreams readFromWriteNode(
+    BlockInputStreams readThroughS3(
         const Context & db_context,
         const SelectQueryInfo & query_info,
         unsigned num_streams);
+    /// helper functions for building the task fetch all data from write node through MPP exchange sender/receiver
+    BlockInputStreams readThroughExchange(unsigned num_streams);
     DM::RNRemoteReadTaskPtr buildDisaggTasks(
         const Context & db_context,
         const DM::ScanContextPtr & scan_context,
@@ -119,7 +121,8 @@ private:
         const std::vector<RemoteTableRange> & remote_table_ranges,
         const pingcap::kv::LabelFilter & label_filter);
     void buildReceiverStreams(const std::vector<RequestAndRegionIDs> & dispatch_reqs, unsigned num_streams, DAGPipeline & pipeline);
-    void filterConditions(NamesAndTypes && source_columns, DAGPipeline & pipeline);
+    void filterConditions(DAGExpressionAnalyzer & analyzer, DAGPipeline & pipeline);
+    void extraCast(DAGExpressionAnalyzer & analyzer, DAGPipeline & pipeline);
     tipb::Executor buildTableScanTiPB();
 
     Context & context;

--- a/dbms/src/Storages/StorageDisaggregatedRemote.cpp
+++ b/dbms/src/Storages/StorageDisaggregatedRemote.cpp
@@ -17,6 +17,7 @@
 #include <Common/ThreadManager.h>
 #include <Common/TiFlashMetrics.h>
 #include <Core/NamesAndTypes.h>
+#include <DataStreams/ExpressionBlockInputStream.h>
 #include <DataStreams/IBlockInputStream.h>
 #include <DataStreams/TiRemoteBlockInputStream.h>
 #include <DataStreams/UnionBlockInputStream.h>
@@ -91,7 +92,7 @@ namespace ErrorCodes
 extern const int DISAGG_ESTABLISH_RETRYABLE_ERROR;
 } // namespace ErrorCodes
 
-BlockInputStreams StorageDisaggregated::readFromWriteNode(
+BlockInputStreams StorageDisaggregated::readThroughS3(
     const Context & db_context,
     const SelectQueryInfo & query_info,
     unsigned num_streams)
@@ -147,8 +148,64 @@ BlockInputStreams StorageDisaggregated::readFromWriteNode(
     // Build InputStream according to the remote segment read tasks
     DAGPipeline pipeline;
     buildRemoteSegmentInputStreams(db_context, remote_read_tasks, query_info, num_streams, pipeline);
-    NamesAndTypes source_columns = genNamesAndTypesForExchangeReceiver(table_scan);
-    filterConditions(std::move(source_columns), pipeline);
+
+    NamesAndTypes source_columns;
+    source_columns.reserve(table_scan.getColumnSize());
+    const auto & remote_segment_stream_header = pipeline.firstStream()->getHeader();
+    for (const auto & col : remote_segment_stream_header)
+    {
+        source_columns.emplace_back(col.name, col.type);
+    }
+
+    analyzer = std::make_unique<DAGExpressionAnalyzer>(std::move(source_columns), context);
+
+    // Handle duration type column
+    std::vector<ExtraCastAfterTSMode> need_cast_column;
+    need_cast_column.reserve(table_scan.getColumnSize());
+    std::unordered_set<ColumnID> col_id_set;
+    for (const auto & expr : table_scan.getPushedDownFilters())
+    {
+        getColumnIDsFromExpr(expr, table_scan.getColumns(), col_id_set);
+    }
+    bool has_need_cast_column = false;
+    for (const auto & col : table_scan.getColumns())
+    {
+        if (col_id_set.contains(col.id))
+        {
+            need_cast_column.push_back(ExtraCastAfterTSMode::None);
+        }
+        else
+        {
+            if (col.id != -1 && col.tp == TiDB::TypeTimestamp)
+            {
+                need_cast_column.push_back(ExtraCastAfterTSMode::AppendTimeZoneCast);
+                has_need_cast_column = true;
+            }
+            else if (col.id != -1 && col.tp == TiDB::TypeTime)
+            {
+                need_cast_column.push_back(ExtraCastAfterTSMode::AppendDurationCast);
+                has_need_cast_column = true;
+            }
+            else
+            {
+                need_cast_column.push_back(ExtraCastAfterTSMode::None);
+            }
+        }
+    }
+    ExpressionActionsChain chain;
+    if (has_need_cast_column && analyzer->appendExtraCastsAfterTS(chain, need_cast_column, table_scan))
+    {
+        ExpressionActionsPtr extra_cast = chain.getLastActions();
+        chain.finalize();
+        chain.clear();
+        for (auto & stream : pipeline.streams)
+        {
+            stream = std::make_shared<ExpressionBlockInputStream>(stream, extra_cast, log->identifier());
+            stream->setExtraInfo("cast after local tableScan");
+        }
+    }
+
+    filterConditions(*analyzer, pipeline);
     return pipeline.streams;
 }
 

--- a/dbms/src/Storages/StorageDisaggregatedRemote.cpp
+++ b/dbms/src/Storages/StorageDisaggregatedRemote.cpp
@@ -160,6 +160,7 @@ BlockInputStreams StorageDisaggregated::readThroughS3(
     analyzer = std::make_unique<DAGExpressionAnalyzer>(std::move(source_columns), context);
 
     // Handle duration type column
+    // If the column is not in the columns of pushed down filter, append a cast to the column.
     std::vector<ExtraCastAfterTSMode> need_cast_column;
     need_cast_column.reserve(table_scan.getColumnSize());
     std::unordered_set<ColumnID> col_id_set;

--- a/dbms/src/Storages/Transaction/TypeMapping.cpp
+++ b/dbms/src/Storages/Transaction/TypeMapping.cpp
@@ -195,6 +195,16 @@ DataTypePtr getDataTypeByColumnInfoForComputingLayer(const ColumnInfo & column_i
     return base;
 }
 
+DataTypePtr getDataTypeByColumnInfoForDisaggregatedStorageLayer(const ColumnInfo & column_info)
+{
+    DataTypePtr base = TypeMapping::instance().getDataType(column_info);
+    if (!column_info.hasNotNullFlag())
+    {
+        return std::make_shared<DataTypeNullable>(base);
+    }
+    return base;
+}
+
 DataTypePtr getDataTypeByFieldType(const tipb::FieldType & field_type)
 {
     ColumnInfo ci = TiDB::fieldTypeToColumnInfo(field_type);

--- a/dbms/src/Storages/Transaction/TypeMapping.h
+++ b/dbms/src/Storages/Transaction/TypeMapping.h
@@ -37,6 +37,8 @@ DataTypePtr getDataTypeByColumnInfoForComputingLayer(const ColumnInfo & column_i
 DataTypePtr getDataTypeByFieldType(const tipb::FieldType & field_type);
 DataTypePtr getDataTypeByFieldTypeForComputingLayer(const tipb::FieldType & field_type);
 
+DataTypePtr getDataTypeByColumnInfoForDisaggregatedStorageLayer(const ColumnInfo & column_info);
+
 TiDB::CodecFlag getCodecFlagByFieldType(const tipb::FieldType & field_type);
 
 // Try best to reverse get TiDB's column info from TiFlash info.

